origin-growth/src/resources/authentication.js

```diff
@@ -0,0 +1,157 @@
+    throw new Error('Recovered and provided accounts do not match')
+  }
+
+  const participant = await db.GrowthParticipant.findOne({
+    where: {
+      ethAddress: accountId.toLowerCase()
+    }
+  })
+
+  if (
+    participant !== null &&
+    participant.status === enums.GrowthParticipantStatuses.Banned
+  ) {
+    logger.warn(`Banned user: ${accountId} tried to enroll`)
+    throw new Error('This user is banned')
+  }
+
+  const authToken =
+    participant === null
+      ? crypto.randomBytes(64).toString('hex')
+      : /* If user uses growth from 2 devices let them share the same auth token.
+         * The caveat is user will need to agree to the terms also on the second
+         * device.
+         */
+        participant.authToken
+
+  const participantData = {
+    ethAddress: accountId.toLowerCase(),
+    status: enums.GrowthParticipantStatuses.Active,
+    agreementId: agreementMessage,
+    authToken: authToken
+  }
+
+  console.log("PARTICIPANT: ", JSON.stringify(participant))
+  if (participant !== null) {
+    console.log("Auth TOKEN: ", authToken, JSON.stringify(participant))
+    await participant.update(participantData)
+    logger.info(`Existing user enrolled into growth campaign: ${accountId}`)
+  } else {
+    await db.GrowthParticipant.create(participantData)
+    logger.info(`New user enrolled into growth campaign: ${accountId}`)
+  }
+
+  await createInviteCode(accountId)
+  return authToken
+}
+
+/**
+ * Fetches user's authentication status
+ * @param {string} token - Growth authentication token
+ * @param {string} accountId - Optional accountIOd parameter
+ *
+ * returns GrowthParticipantAuthenticationStatus
+ *  - Enrolled -> user participates in growth campaign
+ *  - Banned -> user is banned
+ *  - NotEnrolled -> user not a participant yet
+ */
+async function getUserAuthenticationStatus(token, accountId) {
+  const whereFilter = {
+    where: {
+      authToken: token
+    }
+  }
+  if (accountId !== undefined)
+    whereFilter.where.ethAddress = accountId.toLowerCase()
+
+  const growthParticipant = await db.GrowthParticipant.findOne(whereFilter)
+
+  if (growthParticipant === null) {
+    return enums.GrowthParticipantAuthenticationStatus.NotEnrolled
+  } else if (
+    growthParticipant.status === enums.GrowthParticipantStatuses.Banned
+  ) {
+    return enums.GrowthParticipantAuthenticationStatus.Banned
+  } else {
+    return enums.GrowthParticipantAuthenticationStatus.Enrolled
+  }
+}
+
+async function createInviteCode(accountId) {
+  const existingInvite = await db.GrowthInviteCode.findOne({
+    where: {
+      ethAddress: accountId.toLowerCase()
+    }
+  })
+
+  // don't override the existing invite code
+  if (existingInvite !== null) {
+    return
+  }
+
+  const inviteCodesCount = await db.GrowthInviteCode.count({})
+
+  const code =
+    `${accountId.substring(2, 5)}-` +
+    `${accountId.substring(accountId.length - 3, accountId.length)}-` +
+    `${toReadableHex(inviteCodesCount)}`
```

Comment:
Agree that random number encoded to readable string is better. Since I already have a function that converts a number to a readable string I have just generated a random number between 1 and a million like you suggest: 
```
const code =
    `${accountId.substring(2, 5)}` +
    `${accountId.substring(accountId.length - 3, accountId.length)}` +
    `${toReadableHex(Math.round(Math.random() * 1000000))}`
```
Are you ok with this solution?